binding.c

```diff
@@ -11,24 +99,2109 @@ napi_value sn_randombytes_random (napi_env env, napi_callback_info info) {
+  SN_ARGV(2, sodium_memcmp);
+
+  SN_ARGV_TYPEDARRAY(b1, 0)
+  SN_ARGV_TYPEDARRAY(b2, 1)
+
+  SN_THROWS(b1_size != b2_size, "buffers must be of same length")
+
+  SN_RETURN_BOOLEAN(sodium_memcmp(b1_data, b2_data, b1_size))
+}
+
+napi_value sn_sodium_increment(napi_env env, napi_callback_info info) {
+  SN_ARGV(1, sodium_increment);
+  SN_ARGV_TYPEDARRAY(n, 0)
+
+  sodium_increment(n_data, n_size);
+
+  return NULL;
+}
+
+napi_value sn_sodium_add(napi_env env, napi_callback_info info) {
+  SN_ARGV(2, sodium_add);
+
+  SN_ARGV_TYPEDARRAY(a, 0)
+  SN_ARGV_TYPEDARRAY(b, 1)
+
+  SN_THROWS(a_size != b_size, "buffers must be of same length")
+  sodium_add(a_data, b_data, a_size);
+
+  return NULL;
+}
+
+napi_value sn_sodium_sub(napi_env env, napi_callback_info info) {
+  SN_ARGV(2, sodium_sub);
+
+  SN_ARGV_TYPEDARRAY(a, 0)
+  SN_ARGV_TYPEDARRAY(b, 1)
+
+  SN_THROWS(a_size != b_size, "buffers must be of same length")
+  sodium_sub(a_data, b_data, a_size);
+
+  return NULL;
+}
+
+napi_value sn_sodium_compare(napi_env env, napi_callback_info info) {
+  SN_ARGV(2, sodium_compare);
+
+  SN_ARGV_TYPEDARRAY(a, 0)
+  SN_ARGV_TYPEDARRAY(b, 1)
+
+  SN_THROWS(a_size != b_size, "buffers must be of same length")
+  int cmp = sodium_compare(a_data, b_data, a_size);
+
+  napi_value result;
+  napi_create_int32(env, cmp, &result);
+
+  return result;
+}
+
+napi_value sn_sodium_is_zero(napi_env env, napi_callback_info info) {
+  SN_ARGV(1, sodium_is_zero);
+
+  SN_ARGV_TYPEDARRAY(a, 0)
+
+  SN_RETURN_BOOLEAN(sodium_is_zero(a_data, a_size))
+}
+
+napi_value sn_sodium_pad(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, sodium_pad);
+
+  SN_ARGV_TYPEDARRAY(buf, 0)
+  SN_ARGV_UINT32(unpadded_buflen, 1)
+  SN_ARGV_UINT32(blocksize, 2)
+
+  SN_THROWS(unpadded_buflen > buf_size, "unpadded length cannot exceed buffer length")
+  SN_THROWS(blocksize > buf_size, "block size cannot exceed buffer length")
+
+  napi_value result;
+  size_t padded_buflen;
+  sodium_pad(&padded_buflen, buf_data, unpadded_buflen, blocksize, buf_size);
+  assert(napi_create_uint32(env, padded_buflen, &result) == napi_ok);
+  return result;
+}
+
+napi_value sn_sodium_unpad(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, sodium_unpad);
+
+  SN_ARGV_TYPEDARRAY(buf, 0)
+  SN_ARGV_UINT32(padded_buflen, 1)
+  SN_ARGV_UINT32(blocksize, 2)
+
+  SN_THROWS(padded_buflen > buf_size, "unpadded length cannot exceed buffer length")
+  SN_THROWS(blocksize > buf_size, "block size cannot exceed buffer length")
+
+  napi_value result;
+  size_t unpadded_buflen;
+  sodium_unpad(&unpadded_buflen, buf_data, padded_buflen, blocksize);
+  assert(napi_create_uint32(env, unpadded_buflen, &result) == napi_ok);
+  return result;
+}
+
+napi_value sn_crypto_sign_keypair(napi_env env, napi_callback_info info) {
+  SN_ARGV(2, crypto_sign_keypair)
+
+  SN_ARGV_TYPEDARRAY(pk, 0)
+  SN_ARGV_TYPEDARRAY(sk, 1)
+
+  SN_THROWS(pk_size != crypto_sign_PUBLICKEYBYTES, "public key must be 32 bytes")
```

Comment:
can you change the error messages to be "public key must be crypto_sign_PUBLICKEYBYTES bytes"

Reply:
for all the errors, so people look at the constants

Reply:
I think it should be even more precise like: "pk must be crypto_sign_PUBLICKEYBYTES bytes", ie. the "ARG must be CONST bytes"

Reply:
👍

Reply:
fixed by 05082e9